flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java

```diff
@@ -0,0 +1,257 @@
+		this.maxRetentionTime = maxRetentionTime;
+		this.stateCleaningEnabled = minRetentionTime > 1;
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		sortedTimestamps = new LinkedList<>();
+		windowData = new LinkedList<>();
+	}
+
+	@Override
+	public void onProcessingTime(InternalTimer<K, VoidNamespace> timer) throws Exception {
+		if (stateCleaningEnabled) {
+
+			Iterator<Long> keysIt = inputState.keys().iterator();
+			Long lastProcessedTime = lastTriggeringTsState.value();
+			if (lastProcessedTime == null) {
+				lastProcessedTime = 0L;
+			}
+
+			// is data left which has not been processed yet?
+			boolean noRecordsToProcess = true;
+			while (keysIt.hasNext() && noRecordsToProcess) {
+				if (keysIt.next() > lastProcessedTime) {
+					noRecordsToProcess = false;
+				}
+			}
+
+			if (noRecordsToProcess) {
+				inputState.clear();
+				cleanupTsState.clear();
+			} else {
+				// There are records left to process because a watermark has not been received yet.
+				// This would only happen if the input stream has stopped. So we don't need to clean up.
+				// We leave the state as it is and schedule a new cleanup timer
+				registerProcessingCleanupTimer(timerService.currentProcessingTime());
+			}
+		}
+	}
+
+	@Override
+	public void onEventTime(InternalTimer<K, VoidNamespace> timer) throws Exception {
+		long timestamp = timer.getTimestamp();
+		// gets all window data from state for the calculation
+		List<RowData> inputs = inputState.get(timestamp);
+
+		Iterable<Long> keyIter = inputState.keys();
+		for (Long dataTs : keyIter) {
+			insertToSortedList(dataTs);
+		}
+		int index = sortedTimestamps.indexOf(timestamp);
+		for (int i = 0; i < inputs.size(); i++) {
+			forwardedInputQueue.add(inputs.get(i));
+			triggerWindowProcess(inputs, i, index);
+		}
+		windowData.clear();
+		sortedTimestamps.clear();
+	}
+
+	@Override
+	@SuppressWarnings("ConstantConditions")
+	public void emitResult(Tuple2<byte[], Integer> resultTuple) throws Exception {
+		byte[] udafResult = resultTuple.f0;
+		int length = resultTuple.f1;
+		bais.setBuffer(udafResult, 0, length);
+		int rowCount = arrowSerializer.load();
+		for (int i = 0; i < rowCount; i++) {
+			RowData data = arrowSerializer.read(i);
+			RowData key = forwardedInputQueue.poll();
+			reuseJoinedRow.setRowKind(key.getRowKind());
+			rowDataWrapper.collect(reuseJoinedRow.replace(key, data));
+		}
+	}
+
+	void registerProcessingCleanupTimer(long currentTime) throws Exception {
+		if (stateCleaningEnabled) {
+			registerProcessingCleanupTimer(
+				cleanupTsState,
+				currentTime,
+				minRetentionTime,
+				maxRetentionTime,
+				timerService);
+		}
+	}
+
+	void triggerWindowProcess(List<RowData> inputs, int i, int index) throws Exception {
+		int startIndex;
+		int startPos = 0;
+		if (windowData.isEmpty()) {
+			if (i >= lowerBoundary) {
+				for (int j = (int) (i - lowerBoundary); j <= i; j++) {
+					RowData rowData = inputs.get(j);
+					windowData.add(rowData);
```

Comment:
What about just building the windowData list if windowData is empty? I think we can share the same logic about performing the serialization.